DietPi-Software | IPFS Node

nonnorm opened the pull request:
<!--
Before submitting a pull request:
- Please ensure the target branch is "dev" (active development): https://github.com/MichaIng/DietPi/tree/dev
- Please ensure changes have been tested and verified functional.
-->
**Status**: Ready
- [x] Test on VM
- [x] Test on RPi

**Reference**: https://github.com/MichaIng/DietPi/issues/4232

**Commit list/description**:
<!--
- DietPi-Config | Add "Fan control" option to "Performance Options"
![Screenshot](https://xxx.github.com/images/xxx.png)
- DietPi-Config | Add "Fan control" support for Odroid C2
- DietPi-Config | Syntax fix
-->
- DietPi-Software | Add IPFS Node
- DietPi-Software | Add uninstall instructions

nonnorm committed:
v7.1
+ DietPi-Software | Add IPFS Node

nonnorm commented:
Well, I tried updating VirtualBox and now the kernel module is not loading.  Time to start using QEMU.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -9878,6 +9922,42 @@ _EOF_
+
+			Banner_Configuration
+
+			G_EXEC mkdir /mnt/dietpi_userdata/ipfs
```
Use `mkdir -p` or `[[ -d '/mnt/dietpi_userdata/ipfs' ]] ||` to not have a failure when the directory exists already.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -9878,6 +9922,42 @@ _EOF_
+WantedBy=multi-user.target
+_EOF_
+			G_EXEC systemctl daemon-reload
+			G_EXEC systemctl enable ipfs
```
Let's integrate it into `dietpi-services`.

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -9878,6 +9922,42 @@ _EOF_
+{
+	"Addresses.Gateway": "/ip4/127.0.0.1/tcp/8087"
+}
+_EOF_
```
Let's no overwrite existing config files on reinstall, so prefix with `[[ -f '/mnt/dietpi_userdata/ipfs/config' ]]`. I guess the below init could be skipped as well when the config file exists already, so both can be wrapped into the file existence conditional?

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -9878,6 +9922,42 @@ _EOF_
+[Service]
+User=ipfs
+Environment=IPFS_PATH=/mnt/dietpi-userdata/ipfs
+ExecStart=/usr/local/bin/ipfs daemon
```
Btw, not sure in this case, but generally in systemd units processes do not need to be daemonised, since the systemd service itself is. To allow system tracking the process state and catch STDOUT and STDERR for journalctl logging, it's best to run everything in foreground.

But in some cases foreground calls imply an interactive console and such things, which of course would be an overhead or even break the start, so probably it's not applicable here.

nonnorm replied:
I don't ever see the daemon forking into the background, it seems to just stay in the foreground.

MichaIng replied:
Indeed it stays in foreground, so all fine 👍.

MichaIng commented:
I like the idea to merge nodes for distributed projects. Probably the decentralised search engines YaCy and SearX (not added yet) could be added there as well, as the local node as well contributes to the global network.

nonnorm commented:
Coincidentally, I was planning to do SearX sometime soon.

MichaIng commented:
> Well, I tried updating VirtualBox and now the kernel module is not loading. Time to start using QEMU.

Related to IPFS? You mean updating the desktop application or the DietPi guest system?

nonnorm commented:
It was the desktop application, and I hadn't even switched to the IPFS branch yet, so almost certainly not related.

nonnorm commented:
Fixed now (broken permissions on the kext directory).

nonnorm committed:
v7.1
+ Multiple | Fixes

nonnorm committed:
v7.1
+ DietPi-Software | Fix variable call

nonnorm committed:
v7.1
+ DietPi-Software | Add missing !

nonnorm committed:
v7.1
+ DietPi-Software | Add uninstall instructions

nonnorm committed:
v7.1
+ DietPi-Software | Remove systemd service when uninstalling

nonnorm committed:
v7.1
+ DietPi-Software | Fix G_CONFIG_INJECT

nonnorm committed:
v7.1
+ DietPi-Software | Again fix G_CONFIG_INJECT

nonnorm committed:
v7.1
+ DietPi-Software | Fix moved line

nonnorm committed:
v7.1
+ DietPi-Software | Fix userdata path in systemd service

nonnorm committed:
v7.1
+ DietPi-Software | Chown after all files have been created

nonnorm committed:
v7.1
+ DietPi-Software | Make IPFS_PATH an environment variable

nonnorm committed:
v7.1
+ DietPi-Software | Allow access from outside of localhost

nonnorm committed:
v7.1
+ DietPi-Software | Broadcast webui

nonnorm committed:
v7.1
+ DietPi-Software | Remove bashrc file on uninstall

nonnorm committed:
v7.1
+ DietPi-Software | Change YaCy docs link

nonnorm committed:
v7.1
+ DietPi-Software | Fix YaCy software index

nonnorm committed:
v7.1
+ DietPi-Software | Add IPFS user to dietpi group

nonnorm committed:
v7.1
+ DietPi-Software | Chmod the ipfs directory to give dietpi user access

nonnorm committed:
v7.1
+ DietPi-Software | Broadcast gateway

nonnorm commented:
That's nice, the IPFS browser extension lets you select an external API and gateway.  Now I can use IPFS normally as if it was being run on my computer!  Only caveat is that IPFS content won't load on `https://` sites without an SSL certificate on the node.
<img width="318" alt="Screen Shot 2021-04-09 at 4 27 25 PM" src="https://user-images.githubusercontent.com/50060110/114242443-80aa2600-9950-11eb-89bd-aa58c48bdcd5.png">

nonnorm committed:
v7.1

+ DietPi-Software | Change IPFS API port to 5003

nonnorm committed:
v7.1
+ DietPi-Launcher | Change DietPi-NordVPN to DietPi-VPN

nonnorm committed:
Merge branch 'ipfs' of https://github.com/MichaIng/DietPi into ipfs

nonnorm commented:
Currently, I'm testing a mirror of the DietPi (disk) images at http://ipfs.io/ipns/k51qzi5uqu5dit856p8codd8mdkhfgic8m93jl4i2udrr1o35uvtnc2nbvtfyi.

MichaIng commented on dietpi/dietpi-services:
```diff
@@ -194,9 +192,14 @@ Available services:
+			'fahclient'
+			'ipfs'
+			'yacy'
+			'tor'
```
Since the Tor service can also be the Tor HotSpot, which needs to be started before hostapd, please keep it in the network section.

MichaIng committed:
v7.1

+ DietPi-Survey_report | Add IPFS support, newly added with DietPi v7.1

nonnorm committed:
v7.1
+ DietPi-Services | Move tor back to networking section

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -3612,6 +3620,41 @@ _EOF_
+
+			version=$(curl -sSfL 'https://dist.ipfs.io/go-ipfs/versions' | tail -n 1)
+			
+			Download_Install "https://dist.ipfs.io/go-ipfs/$version/go-ipfs_${version}_linux-$arch.tar.gz"
```
`fallback_url` does not work like that. It is only used when `$1` of `Download_Install` is an empty string, not as a general fallback. It is intended for that case the api.github.com "browser_download_url" scrapes return an empty string, because of changed download file names. Here it's similar to api.github.com "tag_name" scrapes, where we only get the version string and build the download URL with it:
```
local version=$(curl -sSfL 'https://dist.ipfs.io/go-ipfs/versions' | tail -n 1)
[[ $version ]] || { version='v0.8.0'; G_DIETPI-NOTIFY 1 "Automatic latest ${aSOFTWARE_NAME[$software_id]} version detection failed. Version \"$version\" will be installed as fallback, but a newer version might be available. Please report this at: https://github.com/MichaIng/DietPi/issues"; }
Download_Install "https://dist.ipfs.io/go-ipfs/$version/go-ipfs_${version}_linux-$arch.tar.gz"
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -9878,6 +9921,46 @@ _EOF_
+
+			fi
+
+			G_EXEC eval "echo -e '#!/bin/dash\nexport IPFS_PATH=/mnt/dietpi_userdata/ipfs' > /etc/bashrc.d/ipfs.sh"
```
Please rename it to `/etc/bashrc.d/dietpi-ipfs.sh`. I aim to have all files that we create, where possible, prefixed with `dietpi-` so that they can be easily identified 🙂.

**EDIT: Another thing**
Is it an issue when the CLI is called with a different user than the `ipfs` service user? If so, we could change that login script to create an alias:
```
alias ipfs='sudo -u ipfs IPFS_PATH=/mnt/dietpi_userdata/ipfs ipfs'
```

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -14536,6 +14619,15 @@ _EOF_
+
+			Banner_Uninstalling
+			systemctl disable --now ipfs
+			rm -r /usr/local/bin/ipfs /mnt/dietpi_userdata/ipfs /etc/systemd/system/ipfs.service /etc/bashrc.d/ipfs.sh
```
Check out other uninstalls where services are involved. To assure that really all related files are removed and avoid error messages when files have been removed before:
```
if [[ -f '/etc/systemd/system/ipfs.service' ]]
then
	systemctl disable --now ipfs
	rm -v /etc/systemd/system/ipfs.service
fi
[[ -d '/etc/systemd/system/ipfs.service.d' ]] && rm -R /etc/systemd/system/ipfs.service.d
rm -Rf /usr/local/bin/ipfs /mnt/dietpi_userdata/ipfs /etc/bashrc.d/ipfs.sh
```

nonnorm committed:
v7.1
+ DietPi-Software | Fixes

MichaIng commented:
Online docs required: https://github.com/MichaIng/DietPi-Docs/issues/425
Did some install tests and it works like a charm 👍.

MichaIng committed:
v7.1

+ DietPi-Software | IPFS: Limit data dir write permissions to IPFS service user, as it's basically a database directory that is not intended to be used directly for move files inside via file servers or such.
+ DietPi-Software | Add some code comments and G_EXEC usage, where applicable

MichaIng commented:
One error during startup:
```
2021/04/11 18:25:50 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size for
details.
```
https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size
```
# sysctl net.core.rmem_max
net.core.rmem_max = 212992
# sysctl -w net.core.rmem_max=2500000
net.core.rmem_max = 2500000
# Issue solved
```
Shall we implement an `/etc/sysctl.d/dietpi-ipfs.conf` to have this applied automatically. Not sure whether ~2.4 MiB packet receive buffer can be an issue for low memory devices?
___________
Another question: I don't see a reason to have the `ipfs` user being member of the `dietpi` group, since its internal data is completely independent from any file system directories, right? The CLI in case needs read permissions, but the `dietpi` group is by default only required for write permissions to the `/mnt/dietpi_userdata` media directories. Shall we skip that group membership, to harden the application a little?

nonnorm committed:
v7.1
+ DietPi-Software | Fix switched around arches

nonnorm commented:
Let's add the file, at least for the sake of mitigating the error.  It seems that the `dietpi` user requires full permissions to the directory for IPFS commands to work, I just did that so that we wouldn't have to apply `777` permissions to the directory.

MichaIng commented:
> It seems that the `dietpi` user requires full permissions to the directory for IPFS commands to work

What has the `dietpi` user to do with it? You mean when one logs in with that user and runs the `ipfs` command from console? For this reason I thought about the CLI alias: https://github.com/MichaIng/DietPi/pull/4236#discussion_r611211633
Generally the `dietpi` user should not be seen as the intended login user and I aim to remove it in favour of an optional custom login user to be created on first boot of DietPi. To have new files and directories in case created with the correct user, so that the web UI remains to have full access, the CLI should be executed as `ipfs` user as well.

nonnorm commented:
Ah.  In that case, yes, let's remove it from the `dietpi` group.

MichaIng commented:
Okay, I'm adding the change, including the CLI alias and the receive buffer raise.

MichaIng committed:
v7.1

+ DietPi-Software | IPFS: Do not add service user to "dietpi" group. Instead, to assure all login users call the correct instance via correct service user, create a CLI alias to invoke the CLI via "sudo -u ipfs".
+ DietPi-Software | IPFS: Raise receive buffer by default to meet the recommendation from the IPFS wiki and mitigate the related error on service start
+ DietPi-Software | IPFS: Add an IPFS node API docs link to the systemd unit

MichaIng committed:
v7.1

+ DietPi-Software | IPFS: Apply increased receive buffer directly on install, to have the error message addressed on first service start already
+ DietPi-Software | Do not call "dietpi-set_hardware headless 0" in every case but only if graphical software is installed

MichaIng committed:
v7.1

+ CHANGELOG | IPFS Node: The InterPlanetary File System peer-to-peer hypermedia protocol node is now available for install

MichaIng commented:
The web interface is unprotected by default. Is there a way at application level to enable login authentication, limit access to local network or such?

MichaIng commented:
I'll merge it, we can further test and think about web UI protection in the issue.
